### PR TITLE
Fix caching linq query with ThenFetchMany

### PR DESCRIPTION
**Pull request by illkostinf**
The commit contains an unit test for #2559 that performs caching linq query with using ThenFetchMany statement and tries to retrieve cached query result set. Starting from nhibernate 5.3 retrieving such cached query result set is not working so unit test is failing. Actually an unit test shows the bug in nhibernate 5.3.

### CODE REVIEW

**Comment by fredericDelaporte** [comment]
Do you mean this was working with 5.2x?

File: src/NHibernate.Test/NHSpecificTest/CachingComplexQuery/FixtureByCode.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+namespace NHibernate.Test.NHSpecificTest.CachingComplexQuery
+{
+	/// <summary>
+	/// Fixture using 'by code' mappings
+	/// </summary>
+	/// <remarks>
+	/// This fixture is identical to <see cref="Fixture" /> except the <see cref="Person" /> mapping is performed 
+	/// by code in the GetMappings method, and does not require the <c>Mappings.hbm.xml</c> file. Use this approach
+	/// if you prefer.
+	/// </remarks>
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/CachingComplexQuery/FixtureByCode.cs:16]: Please delete this comment
```suggestion
```

File: src/NHibernate.Test/NHSpecificTest/CachingComplexQuery/FixtureByCode.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public void TestQueryCachingWithThenFetchMany()
+		{
+			using (var session = OpenSession())
+			{
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/CachingComplexQuery/FixtureByCode.cs:121]: Please stack nested usings:

```csharp
using (var session = OpenSession())
using (var transaction = session.BeginTransaction())
{
```

File: src/NHibernate.Test/NHSpecificTest/CachingComplexQuery/FixtureByCode.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				session.Delete("from Pet");
+				session.Delete("from Child");
+				session.Delete("from Car");
+				session.Delete("from Person");
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/CachingComplexQuery/FixtureByCode.cs:111]: Please delete with `session.CreateQuery("delete from XXX").ExecuteUpdate();`

**Comment by illkostinf** [comment]
> Do you mean this was working with 5.2.x?

Yes, it was working with 5.2.7

**Comment by maca88** [comment]
Rebased to `5.3.x` and added a fix.

File: src/NHibernate/Type/CollectionType.cs [hunk 4]
```diff
@@ -161,7 +161,7 @@ public override object Disassemble(object value, ISessionImplementor session, ob
 		public override void BeforeAssemble(object oid, ISessionImplementor session)
 		{
 			var queryCacheQueue = session.PersistenceContext.BatchFetchQueue.QueryCacheQueue;
-			if (queryCacheQueue == null)
+			if (queryCacheQueue == null || oid == null)
```
Comment (bahusoid) [inline src/NHibernate/Type/CollectionType.cs:164]: Can you please add oid null check before retrieving `queryCacheQueue`

Comment (maca88) [reply src/NHibernate/Type/CollectionType.cs]: Done.

**Review by hazzik — APPROVED** [review]

**Pull request merged**